src/Microsoft.Identity.Web.AgentIdentities/MsalAgentUserIdentityAddIn.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Identity.Abstractions;
+using Microsoft.Identity.Client;
+using Microsoft.Identity.Client.Extensibility;
+
+namespace Microsoft.Identity.Web.AgentIdentities
+{
+    internal static class AgentUserIdentityMsalAddIn
+    {
+        private const string azureAdTokenExchangeScope = "api://AzureADTokenExchange/.default";
+        internal static async Task OnBeforeUserFicForAgentUserIdentityAsync(
+            AcquireTokenByUsernameAndPasswordConfidentialParameterBuilder builder,
+            AcquireTokenOptions? options,
+            ClaimsPrincipal user)
+        {
+            if (options == null || options.ExtraParameters == null)
+            {
+                return;
+            }
+            IServiceProvider serviceProvider = (IServiceProvider)options.ExtraParameters![Constants.ExtensionOptionsServiceProviderKey];
+            options.ExtraParameters.TryGetValue(Constants.AgentIdentityKey, out object? agentIdentityObject);
+            options.ExtraParameters.TryGetValue(Constants.UsernameKey, out object? usernameObject);
+            if (agentIdentityObject is string agentIdentity && usernameObject is string username)
+            {
+                ITokenAcquirerFactory tokenAcquirerFactory = serviceProvider.GetRequiredService<ITokenAcquirerFactory>();
+                IAuthenticationSchemeInformationProvider authenticationSchemeInformationProvider =
+                    serviceProvider.GetRequiredService<IAuthenticationSchemeInformationProvider>();
+
+                ITokenAcquirer tokenAcquirer = tokenAcquirerFactory.GetTokenAcquirer(authenticationSchemeInformationProvider.GetEffectiveAuthenticationScheme(options.AuthenticationOptionsName));
```

Comment:
This tokenAcquirer should be inside the `OnBeforeTokenRequest`, because it does not need to be evaluated if `OnBeforeTokenRequest` is not fired (e.g. when tokens are cached).